### ISSUE TITLE
Suggestion/RFC: Abstain from autocorrection of the Lint/ class cops

### ISSUE THREAD

**Issue by zverok** [rubocop/rubocop#11251]
This is not a "bug report" or "feature request", but rather a request for discussion.

For several big codebases I have worked with, the following was true:
* there was a team/person responsible for setting up Rubocop "our way" (or just introducing it and persuading everybody to make peace with the defaults)
* then, various developers have different patterns of using Rubocop, but there always would be a fair share of those who would just run all the autocorrects possible mindlessly (more considering it autoformatter/style fixer than anything else)

I believe this is true for most of the big organizations using Rubocop. 

In those conditions, autocorrections for some cops are especially dangerous, namely: Lint autocorrections. The reason is: whole point of **linting** is to draw attention to possible errors/wrong semantics, and providing a way to "automatically correct code the way auto-linter would be happy" is defying the point entirely. 

Two recent cops I have in mind, but they just show the principle:
* One is [described here](https://github.com/rubocop/rubocop/issues/11199): `Lint/UnusedBlockArgument` draws attention to an argument that was genuinely _missed_ and _should've been used_, but then, `rubocop -A` conveniently does "so that Rubocop will not complain"
* One is from the recent version changelog `Lint/AssignmentInCondition` cop finds cases like `if foo = bar` where **you could've meant one of two things**: `if foo == bar` or really shortcutting assignment + condition (and should be, for clarity, written like `if (foo = bar)`). The cop is now able to autocorrect to "wrap in parens" everywhere. Which, again, defies the point of the linter! It is totally possible to have a genuine typo (e.g., missing `=`: after 25 years of programming, I had this as recent as last week), but instead of highlighting a possible mistake, Rubocop now strives to mask it.

Now, **why I believe "safe" vs "unsafe" autocorrection marking is not enough**? Because this granularity is too low. Due to extreme flexibility of Ruby, most of autocorrections are "unsafe" (because even something as trivial as replacing `a + a` with `a * 2` depends on definition of `a`), which leads to the fact that people very frequently just do `rubocop -A` because "what's the worst could happen?"

In general, I believe that if many things **could** be autocorrected, it doesn't mean they **should**. 

The only legitimate case for autocorrecting the above mentioned cops I _can_ imagine is "we have a mature, thoroughly-tested codebase where we are just introducing Rubocop; we are 100% sure there are no bugs in codebase, but Rubocop is unhappy with too many 'unused arguments', so let's autocorrect them to continue to use Rubocop wisely."

And I honestly don't believe this case is generic enough (I mean, as a person who did just that several times: introduced Rubocop anew into a living codebase, I can assure that some things I prefer to double-check manually). Even if it is, it might rather suggest a necessity for something like `--apply-initial-fixes` (symmetric to `--auto-gen-config`), which will allow to _once_ perform more aggressive autocorrections, that aren't applicable _in general_.

**Comment by bbatsov** [comment]
> Abstain from autocorrection of the Lint/ class cops

Historically none of them did any autocorrection, as usually there were subtleties that required human input, but we've gradually started to add more and more on that front due to a lot of demand from our users - many people said they were fine with the implications as long as this reduced the manual work for them to some extent. 

> One is from the recent version changelog Lint/AssignmentInCondition cop finds cases like if foo = bar where you could've meant one of two things: if foo == bar or really shortcutting assignment + condition (and should be, for clarity, written like if (foo = bar)). The cop is now able to autocorrect to "wrap in parens" everywhere. Which, again, defies the point of the linter! It is totally possible to have a genuine typo (e.g., missing =: after 25 years of programming, I had this as recent as last week), but instead of highlighting a possible mistake, Rubocop now strives to mask it.

One can make such arguments about pretty much everything in linting - obviously the assumption is that you first saw the warnings and decided to auto-correct them, and reviewed the outcome. Any tool can be used in different ways. :-) But your examples do illustrate my first point - for RuboCop is extremely hard to do know what you meant to do in such cases, that's why I initially wanted to leave those cases to be handled manually. 

Perhaps a good compromise is just to set all (most?) such auto-corrects to be disabled by default, so people would have to explicitly acknowledge they know what they are doing. More documentation probably won't hurt either.

**Comment by dvandersluis** [comment]
As an informational aside (that I mentioned in #11199), putting this in your .rubocop.yml would give you the behaviour you're looking for:

```yaml
Lint:
  AutoCorrect: false
```

I don't think it's inherently bad to have auto correction baked in, even for lints. Personally, when I run RuboCop on a new project, or after upgrading and enabling a bunch of new cops, I prefer to have it autocorrect for me, and then I manually go through the diffs. I accept this as a tradeoff that is useful for me, but because RuboCop is configurable (and in fact, nothing is ever making you run autocorrection at all) someone else can make their own decision (for themselves or their team).

I agree in general that autocorrection safety is not ideal, and we observe many autocorrections trending towards unsafe over time (as false scenarios are reported). It may be that autocorrection needs an implied "Use at your own risk" disclaimer, because as you alluded to, with static analysis, we can never know with perfect certainty that a method call we're looking for is the same method as the cop expects it to be.

**Comment by zverok** [comment]
> Perhaps a good compromise is just to set all (most?) such auto-corrects to be disabled by default, so people would have to explicitly acknowledge they know what they are doing. More documentation probably won't hurt either.

Yes, it is rather a discussion topic (and honestly should've been posted in Discussions, I forgot they exist now :facepalm:).
The disabling by default of autocorrection for everything linting seems sensible, actually!

Another thing that comes to mind is maybe better granularity of "unsafe" to distinguish "generally, it can break the code in an unpredictable way due to semantics and priorities invisible for static analysis" from "this correction should be thoroughly reviewed by human, the sensible correction might be different than the default."

>  putting this in your .rubocop.yml would give you the behaviour you're looking for

Yeah, I am aware (now :facepalm:), but as I am saying, the ticket is more like a general discussion of "shouldn't lint cops be non-autocorrectable by design?"

> I don't think it's inherently bad to have auto correction baked in, even for lints. 

I think I've already expressed it somewhere... One of things I am constantly thinking about is that there is kind of "responsibility" for the default setup/behavior of Rubocop, because it is what the people met with when they first use the tool, including novices, or those who are just trying to introduce the tool. 
And it might be that two types of "unsafety" I am talking about aren't that obvious, and therefore might be confusing and misleading.

**Comment by JunichiIto** [comment]
I submitted a similar issue here: https://github.com/rubocop/rubocop/issues/11322

> there always would be a fair share of those who would just run all the autocorrects possible mindlessly

I strongly agree this line.

If you cannot disable auto-correct for lints by default, perhaps RuboCop should warn users after running lint-related auto correction. For example:

```
$ rubocop -A
Inspecting 4 files
W...

Offenses:

sample.rb:17:20: W: [Corrected] Lint/UnusedMethodArgument: Unused method argument - other. If it's necessary, use _ or _other as an argument name to indicate that it won't be used.
def double(number, other)
                   ^^^^^

4 files inspected, 1 offense detected, 1 offense corrected

WARNING: The following corrections can be incorrect in some contexts, so your review is required:
- sample.rb:17:20: W: [Corrected] Lint/UnusedMethodArgument
```
